src/Management/src/Endpoint/Actuators/HttpExchanges/HttpExchangesDiagnosticObserver.cs

```diff
@@ -0,0 +1,216 @@
+// The .NET Foundation licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Text;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Steeltoe.Management.Diagnostics;
+
+namespace Steeltoe.Management.Endpoint.Actuators.HttpExchanges;
+
+internal sealed class HttpExchangesDiagnosticObserver : DiagnosticObserver, IHttpExchangesRepository
+{
+    private const string DiagnosticName = "Microsoft.AspNetCore";
+    private const string DefaultObserverName = "TraceDiagnosticObserver";
+    private const string StopEventName = "Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop";
+
+    internal const string Redacted = "******";
+
+    private readonly IOptionsMonitor<HttpExchangesEndpointOptions> _optionsMonitor;
+    private readonly ILogger<HttpExchangesDiagnosticObserver> _logger;
+    internal ConcurrentQueue<HttpExchange> Queue { get; } = new();
+
+    public HttpExchangesDiagnosticObserver(IOptionsMonitor<HttpExchangesEndpointOptions> optionsMonitor, ILoggerFactory loggerFactory)
+        : base(DefaultObserverName, DiagnosticName, loggerFactory)
+    {
+        ArgumentNullException.ThrowIfNull(optionsMonitor);
+
+        _optionsMonitor = optionsMonitor;
+        _logger = loggerFactory.CreateLogger<HttpExchangesDiagnosticObserver>();
+    }
+
+    public HttpExchangesResult GetHttpExchanges()
+    {
+        List<HttpExchange> recentExchanges = [];
+        HttpExchangesEndpointOptions options = _optionsMonitor.CurrentValue;
+
+        recentExchanges.AddRange(
+            from exchange in Queue
+            let requestUri = FilterUri(exchange.Request.Uri, options)
+            let filteredRequestHeaders = options.IncludeRequestHeaders ? FilterHeaders(exchange.Request.Headers, options.RequestHeaders) : []
+            let filteredResponseHeaders = options.IncludeResponseHeaders ? FilterHeaders(exchange.Response.Headers, options.ResponseHeaders) : []
+            let principal = options.IncludeUserPrincipal ? exchange.Principal : null
+            let remoteAddress = options.IncludeRemoteAddress ? exchange.Request.RemoteAddress : null
+            let session = options.IncludeSessionId ? exchange.Session : null
+            let request = new HttpExchangeRequest(exchange.Request.Method, requestUri, filteredRequestHeaders, remoteAddress)
+            let response = new HttpExchangeResponse(exchange.Response.Status, filteredResponseHeaders)
+            select new HttpExchange(request, response, exchange.Timestamp, principal, session, exchange.TimeTaken));
+
+        if (options.Reverse)
+        {
+            recentExchanges = recentExchanges.OrderByDescending(exchange => exchange.Timestamp).ToList();
+        }
```

Comment:
```suggestion
        recentExchanges.Sort((first, second) => options.Reverse ? second.Timestamp.CompareTo(first.Timestamp) : first.Timestamp.CompareTo(second.Timestamp));
```

I think we should always sort on timestamp. Multiple threads can add to the queue concurrently, so there's no strong guarantee the queue is ordered by timestamp. Also, using LINQ here copies the list, which is expensive.